hips/hip-9999.md

```diff
@@ -0,0 +1,78 @@
+---
+hip: 9999
+title: "Distributed responsibility for picking"
+authors: [ "Marc Khouzam <marc.khouzam@montreal.ca>" ]
+created: "2021-01-28"
+type: "process"
+status: "draft"
+---
+
+## Abstract
+
+The picking of PRs into a release branch is currently the responsibility of the maintainer
+cutting the release.  It is usually the largest part of cutting a release.  Improving
+the picking process would make cutting releases easier with a reduced risk of error.
+
+## Motivation
+
+In the current picking process the maintainer that merges a PR is only required to tag that PR with a `needs-pick` 
+label. The day of the patch release the maintainer cutting the release will go through each PR
+with a `needs-pick` label and manually cherry-pick it into the release branch.  Once the
+cherry-pick is done, the `picked` label is added to the PR.
+
+There are multiple issues with this process:
+1. Manual work for the maintainer doing the release to pick each relevant PR.
+1. Possible merge conflicts during the cherry-picks as the code on master may
+have substantially diverged from the code in the release branch.  This is particularly concerning
+as the maintainer cutting the release may not be the one who reviewed the PR in question
+and may not be familiar with its details, thus making resolving the conflicts difficult
+and error prone.
+1. There is basically no soak time for the cherry-picks or merge conflict resolution code:
+some quick manual testing and the unit tests are the only tests these changes go through. 
+
+## Rationale
+
+By cherry-picking PRs as soon as they are merged into the master branch will allow to
+address the issues mentioned above.  It will reduce the manual work required
+at release time, will reduce the risk of errors during merge conflicts, and will provide
+more time to soak the release branch content.
+
+## Specification
+
+The pick process would be modified to the following:
+1. Once a PR is merged into the master branch the `needs-pick` label will still be applied.
+This is to reduce the risk of forgetting to cherry-pick a PR.
+1. Once the PR is merged into the master branch, it becomes the responsibility of the merging maintainer to
+cherry-pick it into the release branch; this should be done through a new PR and be done as
```

Comment:
The goal of a PR process is to review the code before it's merged into the main branch. In this situation, we're not asking for a review; we're just trying to move what was already approved into a staging branch.

Why should the commit go through the PR process a second time? Can't the maintainer just cherry-pick the commits into the release branch as we do today?

Reply:
You can also see this process ballooning wildly out of control if we consider maintaining more than one patch release branch as we've discussed in the past. The PR queue would explode exponentially. I would not want to look through a PR queue that contained the same set of commits three times for three separate branches, all requiring the same review.

Reply:
The first goal of this HIP is to lighten the burden of the maintainer doing the release, so if any of the steps actually make the effort bigger, I absolutely agree it should be changed.  I'll explain my train of thought and we'll adapt the proposal to Helm's reality.

But first let me just comment on
> The goal of a PR process is to review the code before it's merged into the main branch.
> It's more about knowledge transfer than code quality.

This is an important point that I was not aware of.  I am a firm believer that code reviews improve code quality, and I always assumed that was the reason behind PR reviews.  But I must admit that I've often struggled with finding the right balance between the large amount of time a thorough review requires, and the value that investment actually adds. (I would love to discuss this someday over a beer.)  So, with the different perspective you bring up, some of my thoughts on this HIP may not be justified.

Ok, back to my original thoughts when I wrote the HIP.
I immediately realize that I should not have implied going through the full PR process with potentially multiple reviewers, that was a mistake.  What I had in mind was simply using a PR instead of a direct commit.  The benefits I saw are:
1. easier to track and query PRs to know what was included in the patch release or was not
2. for the rare case of substantial merge conflicts, a PR allows to request an extra reviewer (for code quality)
3. allows a maintainer, if desired, to ask the original contributor to submit the PR

But a cherry-pick that does not have any conflicts could simply be created as a PR and merged immediately by the maintainer.

That said, I'll update the HIP to drop the point of using PRs.